libraries/AP_DDS/AP_DDS_Client.cpp

```diff
@@ -0,0 +1,245 @@
+    const char* participant_ref = "participant_profile";
+    const auto participant_req_id = uxr_buffer_create_participant_ref(&session, reliable_out, participant_id,0,participant_ref,UXR_REPLACE);
+
+    // Topic
+    const uxrObjectId topic_id = {
+        .id = topics[0].topic_id,
+        .type = UXR_TOPIC_ID
+    };
+    const char* topic_ref = topics[0].topic_profile_label;
+    const auto topic_req_id = uxr_buffer_create_topic_ref(&session,reliable_out,topic_id,participant_id,topic_ref,UXR_REPLACE);
+
+    // Publisher
+    const uxrObjectId pub_id = {
+        .id = topics[0].pub_id,
+        .type = UXR_PUBLISHER_ID
+    };
+    const char* pub_xml = "";
+    const auto pub_req_id = uxr_buffer_create_publisher_xml(&session,reliable_out,pub_id,participant_id,pub_xml,UXR_REPLACE);
+
+    // Data Writer
+    const char* data_writer_ref = topics[0].dw_profile_label;
+    const auto dwriter_req_id = uxr_buffer_create_datawriter_ref(&session,reliable_out,dwriter_id,pub_id,data_writer_ref,UXR_REPLACE);
+
+    //Status requests
+    constexpr uint8_t nRequests = 4;
+    const uint16_t requests[nRequests] = {participant_req_id, topic_req_id, pub_req_id, dwriter_req_id};
+
+    constexpr int maxTimeMsPerRequestMs = 250;
+    constexpr int requestTimeoutMs = nRequests * maxTimeMsPerRequestMs;
+    uint8_t status[nRequests];
+    if (!uxr_run_session_until_all_status(&session, requestTimeoutMs, requests, status, nRequests)) {
+        // TODO add a failure log message sharing the status results
+        return false;
+    }
+
+    return true;
+}
+
+void AP_DDS_Client::write()
+{
+    WITH_SEMAPHORE(csem);
+    if (connected) {
+        ucdrBuffer ub;
+        uint32_t topic_size = builtin_interfaces_msg_Time_size_of_topic(&time_topic, 0);
+        uxr_prepare_output_stream(&session,reliable_out,dwriter_id,&ub,topic_size);
+        const bool success = builtin_interfaces_msg_Time_serialize_topic(&ub, &time_topic);
+        if (!success) {
+            // TODO sometimes serialization fails on bootup. Determine why.
+            // AP_HAL::panic("FATAL: DDS_Client failed to serialize\n");
+        }
+
+    }
+
+}
+
+void AP_DDS_Client::update()
+{
+    WITH_SEMAPHORE(csem);
+
+    update_topic(&time_topic);
+    write();
+    connected = uxr_run_session_time(&session, 1);
+}
+
+/*
+  implement C functions for serial transport
+ */
+extern "C" {
+#include <uxr/client/profile/transport/serial/serial_transport_platform.h>
+}
+
+bool uxr_init_serial_platform(void* args, int fd, uint8_t remote_addr, uint8_t local_addr)
+{
+    return true;
+}
+
+bool uxr_close_serial_platform(void* args)
+{
+    return true;
+}
+
+size_t uxr_write_serial_data_platform(void* args, const uint8_t* buf, size_t len, uint8_t* errcode)
+{
+    if (dds_port == nullptr) {
+        *errcode = 1;
+        return 0;
+    }
+    ssize_t bytes_written = dds_port->write(buf, (size_t)len);
+    if (bytes_written <= 0) {
+        *errcode = 1;
+        return 0;
+    }
+    *errcode = 0;
```

Comment:
populate errcode to something sensible (is there an enum?)

Reply:
Handled in df8e77bdbc6de6288d8c9d4e85979fe1c0ea5797. It's now marked as a Doxygen todo, and can be solved at a later date after have discussions with EProsima on weak pointers to the `uxr_*` functions we redefined.